### DISCUSSION TITLE
The size of the proposed random crop ROI is larger than the image size

### DISCUSSION THREAD

**Discussion by ramchandracheke** [Project-MONAI/MONAI#2256] [General]
Hi,

Thank you for this wonderful project.

File "/home/ramchand/.local/lib/python3.7/site-packages/monai/transforms/utils.py", line 334, in generate_pos_neg_label_crop_centers
raise ValueError("The size of the proposed random crop ROI is larger than the image size.")

ValueError: The size of the proposed random crop ROI is larger than the image size.
RuntimeError: applying transform <monai.transforms.croppad.dictionary.RandCropByPosNegLabeld object at 0x7ffecef93550>

Could you please help me to solve this error. as My image size is variable(no of slices are different in each nifti file.eg. (342, 342, 91) and I am using Unet. (128,128,128). I am using following combination of transform.

```python
train_transforms = Compose(
[
LoadImaged(keys=["image", "label"]),
AddChanneld(keys=["image", "label"]),
Spacingd(keys=["image", "label"], pixdim=(
1.5, 1.5, 1.0), mode=("bilinear", "nearest")),
Orientationd(keys=["image", "label"], axcodes="RAS"),
ScaleIntensityRanged(
keys=["image"], a_min=-400, a_max=1500,
b_min=0.0, b_max=1.0, clip=True,
),
CropForegroundd(keys=["image", "label"], source_key="image"),
RandCropByPosNegLabeld(
keys=["image", "label"],
label_key="label",
spatial_size=(128,128,128),
pos=1,
neg=1,
num_samples=4,
image_key="image",
image_threshold=0,
),ToTensord(keys=["image", "label"]),
]
)
```

Thanks a million!

**Comment by rijobro** [comment]
Hi @ramchandracheke, your example image had 91 slices in the z-dimension. This is already smaller than the 128 that you've requested of `RandCropByPosNegLabeld`. You're also using `Spacing` and `CropForegroundd` before calling `RandCropByPosNegLabeld`, so it's hard to know what size your images are before `RandCropByPosNegLabeld`.

To debug this, why don't you comment out the final `RandCropByPosNegLabeld` transform. You can then run through all of your data and check the size. Something like (completely untested):

```python
min_sizes = (10000, 10000, 10000)
for d in dataset:
    shape = d["image"].shape[1:]
    for i in range(len(shape)):
        if shape[i] < min_sizes[i]:
            min_sizes[i] = shape[i]
print(min_sizes)
```

**Reply by ramchandracheke** [reply]
Hi @rijobro,
Yeah. Thank you for your reply. My data-set consist of different volumes with different number of slices. Therefore, I found that the volume with minimum dimension of (512,512,74) to (512,512,974). Yes you spotted right that, the dimensions of the image is reduced further after using CropForegroundd. Therefore, I am just using below transform.  

```python
train_transforms = Compose([
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image", "label"]),
        ScaleIntensityRanged(
            keys=["image"], a_min=-400, a_max=1500,
            b_min=0.0, b_max=1.0, clip=True,
        ),
        RandSpatialCropd(keys=["image", "label"], roi_size=(128, 128,128), random_size=False),
        SpatialPadd(keys=["image", "label"],spatial_size=(128,128,128), mode ='reflect'),

        ToTensord(keys=["image", "label"]),
    ]
)

```

This is working, however I think using RandCropByPosNegLabeld I could get better segmentation.

Thank you very much!

**Reply by rijobro** [reply]
If you want to use `RandCropByPosNegLabeld`, could you do `SpatialPadd` first, to make sure that everything is at least `(128, 128, 128)`, and then use `RandCropByPosNegLabeld`?

By the way, I think in your current set up, calling `SpatialPadd` after `RandSpatialCropd` is redundant, since your images will already be `(128, 128, 128)` after `RandSpatialCropd`, so `SpatialPadd` won't do anything.

**Reply by ramchandracheke** [reply]
Hi @rijobro, 

Thanks for your inputs. I changed the order to --> `CropForegroundd`, -->`RandCropByPosNegLabeld` and --> `Rand3DElasticd` and it works for me. I observed that when I added `Rand3DElasticd` the loss is almost constant loss where as its decreasing steadily without using using Rand3DElasticd augmentation. In segmentation task its elastic deformation is used widely as a data augmentation. However, in this case the network converging very slow. Do you have any suggestion on this?

Thanks a million!

**Reply by rijobro** [reply]
Would be a good idea to visualise your images after the elastic transformation. If you get the parameters wrong, the resulting image can look very different compared to the original.

**Reply by ramchandracheke** [reply]
Yeah... That's true!
